libs/langchain-community/src/chat_models/tencent_hunyuan.ts

```diff
@@ -0,0 +1,495 @@
+              if (data?.Response?.Error?.Message) {
+                error = data?.Response?.Error?.Message;
+                requestId = data?.Response?.RequestId;
+              } else if (data?.ErrorMsg?.Message) {
+                // handle streaming error
+                error = data?.ErrorMsg?.Message;
+                requestId = data?.Id;
+              }
+
+              if (error) {
+                if (rejected) {
+                  return;
+                }
+                rejected = true;
+                reject(new Error(`[${requestId}] ${error}`));
+                return;
+              }
+
+              const { Content } = data.Choices[0].Delta;
+              const { FinishReason } = data.Choices[0];
+
+              // on the first message set the response properties
+              if (!response) {
+                response = {
+                  Id: data.Id,
+                  Created: data.Created,
+                  Usage: data.Usage,
+                  Note: data.Note,
+                  RequestId: data.Id,
+                  Choices: [
+                    {
+                      ...data.Choices[0],
+                      Message: data.Choices[0].Delta,
+                    },
+                  ],
+                };
+              } else {
+                response.Usage = data.Usage;
+                response.Choices[0].Message.Content += Content;
+              }
+
+              void runManager?.handleLLMNewToken(Content ?? "");
+              if (FinishReason) {
+                if (resolved || rejected) {
+                  return;
+                }
+                resolved = true;
+                resolve(response);
+              }
+            }
+          ).catch((error) => {
+            if (!rejected) {
+              rejected = true;
+              reject(error);
+            }
+          });
+        })
+      : await this.completionWithRetry(
+          {
+            ...params,
+            Messages: messagesMapped,
+          },
+          false,
+          options?.signal
+        ).then<ChatCompletionResponse>((data) => {
+          const response = data?.Response;
+          if (response?.Error?.Message) {
+            throw new Error(
+              `[${response.RequestId}] ${response.Error.Message}`
+            );
+          }
+          return response;
+        });
+
+    const {
+      TotalTokens = 0,
+      PromptTokens = 0,
+      CompletionTokens = 0,
+    } = data.Usage;
+
+    const text = data.Choices[0]?.Message?.Content ?? "";
+    const generations: ChatGeneration[] = [];
+    generations.push({
+      text,
+      message: new AIMessage(text),
+    });
+
+    return {
+      generations,
+      llmOutput: {
+        tokenUsage: {
+          totalTokens: TotalTokens,
+          promptTokens: PromptTokens,
+          completionTokens: CompletionTokens,
+        },
+      },
+    };
+  }
+
+  /** @ignore */
+  async completionWithRetry(
```

Comment:
nit: we generally prefer these streaming implementations to be generators, but if it works, that's fine

Reply:
I have refactor the stream part of the chat model, with `_streamResponseChunks` being implemented, which returns `AsyncGenerator<ChatGenerationChunk>`. In `_generate`, it calls `_streamResponseChunks` when streaming is true, otherwise it calls `completionWithRetry`.